components/eamxx/src/diagnostics/pbl_entrainment_budget.cpp

```diff
@@ -0,0 +1,332 @@
+        const auto sd_icol = ekat::subview(sd_v, icol);
+        const auto su_icol = ekat::subview(su_v, icol);
+        const auto ld_icol = ekat::subview(ld_v, icol);
+        const auto lu_icol = ekat::subview(lu_v, icol);
+
+        // tracked
+        const auto qt_icol = ekat::subview(qt_v, icol);
+        const auto tl_icol = ekat::subview(tl_v, icol);
+
+        auto prev_qtot_icol = ekat::subview(prev_qtot_v, icol);
+        auto prev_tliq_icol = ekat::subview(prev_tliq_v, icol);
+
+        auto ws = wsm.get_workspace(team);
+        ekat::Unmanaged<WSMgr::view_1d<Real>> tm_grad;
+        ws.take_many_contiguous_unsafe<wsms>({"tm_grad"}, {&tm_grad});
+
+        // We first want to find the PBL inversion. There are three methods to
+        // do so. All our methods here rely on the *gradient* of state fields
+        // (for now). First, we can simply find the first level from the surface
+        // that has a a temperature "inversion" (temperature goes positive
+        // instead of negative). Second, we can find the level which has the
+        // biggest positive jump in theta_l. Third, we can find the level which
+        // has the biggest negative jump in qt.
+
+        int opt_tm_grad_lev = 1;
+        // Find tm_grad (tm_grad is a catch-all for the 3 methods)
+        Kokkos::parallel_for(
+            Kokkos::TeamVectorRange(team, 1, num_levs), [&](int k) {
+              auto pm_diff = pm_icol(k - 1) - pm_icol(k);
+              if(pblinvalg == 1) {
+                // pblinvalg = 1 ---> based solely on
+                // d(T_mid)/d(p_mid); finding the min (because
+                // d(p_mid) < 0), so keeping signs
+                tm_grad(k) = (tm_icol(k - 1) - tm_icol(k)) / pm_diff;
+              } else if(pblinvalg == 2) {
+                // pblinvalg = 2 ---> based solely on
+                // d(theta_l)/d(p_mid); finding the min
+                // (because d(p_mid) < 0), so keeping signs
+                tm_grad(k) = (tl_icol(k - 1) - tl_icol(k)) / pm_diff;
+              } else if(pblinvalg == 3) {
+                // pblinvalg = 3 ---> based solely on
+                // d(q_t)/d(p_mid); finding the max (because
+                // d(p_mid) < 0), so reversing signs
+                tm_grad(k) = -(qt_icol(k - 1) - qt_icol(k)) / pm_diff;
+              }
+            });
+        team.team_barrier();
+
+        // Find minimum gradient, because d(p_mid) < 0 in definition above
+        // Starting from the surface, and ensuring p_mid > 70000.0 Pa,
+        // to avoid resolving to some odd place higher up in the atmosphere.
+        using minloc_t       = Kokkos::MinLoc<Real, int>;
+        using minloc_value_t = typename minloc_t::value_type;
+        minloc_value_t minloc;
+        Kokkos::parallel_reduce(
+            Kokkos::TeamVectorRange(team, 1, num_levs),
+            [&](const int &k, minloc_value_t &result) {
+              if(tm_grad(k) < result.val && pm_icol(k) > 70000.0) {
```

Comment:
You never check that `tm_grad(k)<0`. I guess this usually happens at some point in the v direction, but as a software dev with little atm modeling knowlege I wonder "what if it never occurs?"